### PR TITLE
COM-392 Appts: Start and end time not mandatory

### PR DESCRIPTION
**Pull request by prashantsable1991**
Removed required from start and end time. Added logic to save an appointment when start time and end time, not provided.

### CODE REVIEW

File: ui/app/appointments/controllers/manage/calendar/appointmentsCalendarViewController.js [hunk 1]
```diff
@@ -117,14 +117,15 @@ angular.module('bahmni.appointments')
...
             }, true);
-
             var setAppointments = function (params) {
                 autoRefreshStatus = false;
                 return appointmentsService.getAllAppointments(params).then(function (response) {
                     $scope.allAppointmentsForDay = response.data;
                     var filteredAppointments = appointmentsFilter($scope.allAppointmentsForDay, $state.params.filterParams);
                     $rootScope.appointmentsData = filteredAppointments;
                     autoRefreshStatus = true;
+                    $rootScope.appointmentStartTime = _.last($scope.allAppointmentsForDay);
```
Comment (lauravignoli) [inline ui/app/appointments/controllers/manage/calendar/appointmentsCalendarViewController.js]: @prashantsable1991 Two comments:

1- It should find the earliest slot, not the one available after the last appointment of the day. 
2- It should allow for appointments at the same time if they are for different services.

**Comment by lauravignoli** [comment]
@prashantsable1991 Great work!!!

**Comment by tmvumbi2** [comment]
@rubailly I tried adding a new appointment from an empty calendar and got the error below. This error disappears if there exists an appointment (where the time slot has been manually populated)
image

**Comment by tmvumbi2** [comment]
@rubailly I noticed the bugs mentioned on Laura's comment.
Bug 1:
- I created an appointment on 9 July (from 10:30 to 11 am)
- I tried creating a new appointment on 10 Jul (where there is no appointment scheduled yet); I didn't specify the timeslot and the system selected 11 to 11.30 am (instead of picking the earliest slot)

Bug 2:
- I tried creating an appointment with a different service, and the system booked the next available spot. Expected result: the system should allow appointment at the same time if the service is different (or if the provider is different)

File: ui/app/appointments/controllers/manage/appointmentsCreateController.js [hunk 2]
```diff
@@ -93,14 +93,49 @@ angular.module('bahmni.appointments')
                 }
             };
 
+            var setAppointmentsEndTimeAmPm = function (date) {
```
Comment (tmvumbi2) [inline ui/app/appointments/controllers/manage/appointmentsCreateController.js]: @rubailly I'd suggest we name this method `calculateAppointmentEndTime`

File: ui/app/appointments/controllers/manage/appointmentsCreateController.js [hunk 3]
```diff
@@ -93,14 +93,49 @@ angular.module('bahmni.appointments')
...
+                var strTime = moment(date).add(30, 'm').toDate();
```
Comment (tmvumbi2) [inline ui/app/appointments/controllers/manage/appointmentsCreateController.js]: @rubailly '30' in here corresponds to the appointment duration? I'm not sure if the value should be hard-coded; could you please check if there is a variable storing the minimum duration that we can use here (like `$scope.minDuration`)?

File: ui/app/appointments/controllers/manage/appointmentsCreateController.js [hunk 4]
```diff
@@ -505,6 +540,7 @@ angular.module('bahmni.appointments')
             };
 
             var saveAppointment = function (appointment) {
+                console.log(appointment.startDateTime, 'starttime');
```
Comment (tmvumbi2) [inline ui/app/appointments/controllers/manage/appointmentsCreateController.js]: @rubailly please remove the debug `console.log`

**Comment by rubailly** [comment]
@tmvumbi2 It looks like it can't find the earliest slot. I created two appointments for the same service, one at 8am and another at 2pm. I then created a third one without specifying the start time and end time. -It went to 2:30 pm 

Screenshot from 2019-07-16 15-21-28

**Comment by rubailly** [comment]
Screenshot from 2019-07-18 13-16-49
Screenshot from 2019-07-18 13-18-39

@tmvumbi2 The way it is working now, if you don't specify the appointment start and end time, it will find the earliest appointment and use the same starting and end time. So you end up with multiple appointments that start and end at the same time for the same service and provider.

**Comment by rubailly** [comment]
@tmvumbi2 -All is working well now. The remaining thing is the logic to handle the edge cases when the day has been fully booked. 1) You have to click the save button twice to see the alert that says all slots have been booked. Ideally, you should see the alert automatically after clicking the save button. 2) When you force one appointment to go on a used slot, all the other appointments you create and you don't
 specify the start/end time, they will mix with the appointment you created previously.
Screenshot from 2019-07-19 13-50-07

I, however, think that this is not in the scope of this story, I'll merge this and ping Lisa to create a separate story to handle this logic.

**Pull request merged**